### PR TITLE
[improvement](binlog)Support inverted index format v2 in CCR

### PR DESCRIPTION
**Pull request by qidaye**
## Proposed changes

1. Support inverted index format v2
2. Add regression cases for index format v1 & v2
3. Fix some bugs in Syncer functions

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by qidaye** [comment]
run buildall

File: regression-test/suites/ccr_syncer_p1/inverted_index/test_backup_restore.groovy [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+        sql """ 
+                BACKUP SNAPSHOT ${context.dbName}.${snapshotName} 
+                TO `__keep_on_local__` 
+                ON (${tableName})
+                PROPERTIES ("type" = "full")
+            """
+        while (syncer.checkSnapshotFinish() == false) {
+            Thread.sleep(3000)
+        }
```
Comment (w41ter) [inline regression-test/suites/ccr_syncer_p1/inverted_index/test_backup_restore.groovy:165]:
```suggestion
        syncer.waitAllSnapshotFinish()
```

File: regression-test/suites/ccr_syncer_p1/inverted_index/test_backup_restore.groovy [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        assertTrue(syncer.getSnapshot("${snapshotName}", "${tableName}"))
+        assertTrue(syncer.restoreSnapshot(true))
+        while (syncer.checkRestoreFinish() == false) {
+            Thread.sleep(3000)
+        }
```
Comment (w41ter) [inline regression-test/suites/ccr_syncer_p1/inverted_index/test_backup_restore.groovy:170]:
```suggestion
        syncer.waitTargetRestoreFinish()
```

File: be/src/service/backend_service.cpp [hunk 3]
```diff
@@ -307,41 +307,81 @@ void _ingest_binlog(StorageEngine& engine, IngestBinlogArg* arg) {
...
+                                 << get_segment_index_file_size_url
+                                 << ", status=" << status.to_string();
+                    status.to_thrift(&tstatus);
+                    return;
+                }
 
-            segment_index_file_sizes.push_back(segment_index_file_size);
-            segment_index_file_urls.push_back(std::move(get_segment_index_file_size_url));
+                segment_index_file_sizes.push_back(segment_index_file_size);
+                segment_index_file_urls.push_back(std::move(get_segment_index_file_size_url));
```
Comment (w41ter) [inline be/src/service/backend_service.cpp:383]: It looks like the only difference is the url and index file name, how about changing it to the following code?

```c++
auto get_segment_index_file_size_url = (tablet_schema->get_reverted_index_storage_format() == V1) ? fmt::format(xxxx, index_id) : fmt::format(xxxx, -1);
```

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -2411,14 +2411,24 @@ std::string Tablet::get_segment_filepath(std::string_view rowset_id, int64_t seg
...
 
 std::string Tablet::get_segment_index_filepath(std::string_view rowset_id, int64_t segment_index,
                                                int64_t index_id) const {
-    // TODO(qiye): support inverted index file format v2, when https://github.com/apache/doris/pull/30145 is merged
-    return fmt::format("{}/_binlog/{}_{}_{}.idx", _tablet_path, rowset_id, segment_index, index_id);
+    auto format = _tablet_meta->tablet_schema()->get_inverted_index_storage_format();
+    if (format == doris::InvertedIndexStorageFormatPB::V1) {
+        return fmt::format("{}/_binlog/{}_{}_{}.idx", _tablet_path, rowset_id, segment_index,
+                           index_id);
+    } else {
```
Comment (w41ter) [inline be/src/olap/tablet.cpp:2429]: Consider adding a check:

```c++
DCHECK(index_id == -1);
```

**Review by w41ter — CHANGES_REQUESTED** [review]
Reset LGTM

File: be/src/service/backend_service.cpp (hunk 3 above)
Comment (qidaye) [reply be/src/service/backend_service.cpp:383]: V1 is separated index files by column, V2 is one compound file for all columns. So the logic can not merge together.

**Comment by qidaye** [comment]
run buildall

**Review by airborne12 — APPROVED** [review]
LGTM

**Review by w41ter — APPROVED** [review]

**Comment by qidaye** [comment]
run p0

**Comment by qidaye** [comment]
run p1

**Comment by qidaye** [comment]
run buildall

**Comment by qidaye** [comment]
run p0

**Comment by qidaye** [comment]
run p0

**Comment by qidaye** [comment]
run p0

**Comment by qidaye** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**